ryoanji/src/ryoanji/nbody/traversal_ewald_cpu.hpp

```diff
@@ -0,0 +1,600 @@
+        }
+
+ewald_shells_end:
+
+//ewald_waves_begin:
+
+        if (gravity_switch & GRAV_NO_EWALD_KSPACE) goto ewald_waves_end;
+
+        for (auto ed : ed_list)
+        {
+            auto hdotx = ed.hx*dx + ed.hy*dy + ed.hz*dz;
+            auto c = std::cos(hdotx);
+            auto s = std::sin(hdotx);
+            auto cs_diff = ed.hfac_cos*s - ed.hfac_sin*c;
+            auto cs_sum  = ed.hfac_cos*s + ed.hfac_sin*c;
+
+            ax_t    += ed.hx * cs_diff;
+            ay_t    += ed.hy * cs_diff;
+            az_t    += ed.hz * cs_diff;
+            ugrav_t -=         cs_sum;
+        }
+ewald_waves_end:
+
+        ax[j]    += ax_t;
+        ay[j]    += ay_t;
+        az[j]    += az_t;
+        ugrav[j] += ugrav_t;
+    }
+}
+
+/*! @brief computes gravitational acceleration for all particles in the specified group
+ *
+ * @tparam T1                   float or double
+ * @tparam T2                   float or double
+ * @param[in]    groupIdx       leaf cell index in [0:octree.numLeafNodes()] to compute accelerations for
+ * @param[in]    childOffsets   child node index of each node
+ * @param[in]    internalToLeaf map to convert an octree node index into a cstone leaf index
+ * @param[in]    multipoles     array of length @p octree.numTreeNodes() with the multipole moments for all nodes
+ * @param[in]    layout         array of length @p octree.numLeafNodes()+1, layout[i] is the start offset
+ *                              into the x,y,z,m arrays for the leaf node with index i. The last element
+ *                              is equal to the length of the x,y,z,m arrays.
+ * @param[in]    x              x-coordinates
+ * @param[in]    y              y-coordinates
+ * @param[in]    z              z-coordinates
+ * @param[in]    h              smoothing lengths
+ * @param[in]    m              masses
+ * @param[in]    box            global coordinate bounding box
+ * @param[in]    theta          accuracy parameter
+ * @param[in]    G              gravitational constant
+ * @param[inout] ax             location to add x-acceleration to
+ * @param[inout] ay             location to add y-acceleration to
+ * @param[inout] az             location to add z-acceleration to
+ * @param[inout] ugrav          location to add gravitational potential to
+ *
+ * Note: acceleration output is added to destination
+ */
+template<class MType, class T1, class T2, class Tm>
+void computeGravityGroupReplica(TreeNodeIndex groupIdx, const TreeNodeIndex* childOffsets, const TreeNodeIndex* internalToLeaf,
+                                const cstone::SourceCenterType<T1>* centers, MType* multipoles, const LocalIndex* layout,
+                                const T1 dx_target, const T1 dy_target, const T1 dz_target,
+                                const T1* x, const T1* y, const T1* z, const T2* h, const Tm* m, float G, T1* ax, T1* ay,
+                                T1* az, T1* ugrav)
+{
```

Comment:
If we change the signature here to directly take `targetCenter` and `targetSize`, also for the existing variant of this function, and compute targetCenter and targetSize at the call site, we only need one version of this function to traverse both the center box and replicas.

Reply:
computeGravityGroupReplica() differs from computeGravityGroup() in more ways than just the value of targetCenter. First, the P2P loop assumes that all particles in the target leaf interact with the particles in the replica leaf. This is different than the normal case where the target must not be its own source. Second, I need to apply {dx,dy,dz}_target to each target particle (in both P2P and M2P). One could pass a new array of target x,y,z values, but then a new function would still be required and would, I feel, overcomplicate the calling code.